codex/stores/maintenance.nim

```diff
@@ -10,92 +10,323 @@
   )
 
-  without iter =? expirations, err:
-    trace "Unable to obtain blockExpirations iterator from repoStore"
-    return
+proc deleteBatch(self: DatasetMaintainer, treeCid: Cid, datasetMd: DatasetMetadata): Future[?!void] {.async.} =
+  var datasetMd = datasetMd
+
+  datasetMd.checkpoint.timestamp = self.clock.now
+
+  if err =? (await self.recordCheckpoint(treeCid, datasetMd)).errorOption:
+    return failure(err)
+
+  logScope:
+    treeCid = treeCid
+    manifestsCids = datasetMd.manifestsCids
+    startingIndex = datasetMd.checkpoint.progress
+
+  trace "Deleting a batch of blocks", size = self.batchSize
+
+  var index = datasetMd.checkpoint.progress
+  while (index < datasetMd.checkpoint.progress + self.batchSize) and
+          (index < datasetMd.leavesCount):
+    if err =? (await self.blockStore.delBlock(treeCid, index)).errorOption:
+      error "Error deleting a block", msg = err.msg, index
+
+    index.inc
+
+    await sleepAsync(50.millis) # cooperative scheduling
+
+  if index >= datasetMd.leavesCount:
+    trace "All blocks deleted from a dataset", leavesCount = datasetMd.leavesCount
+
+    for manifestCid in datasetMd.manifestsCids:
+      if err =? (await self.blockStore.delBlock(manifestCid)).errorOption:
+        error "Error deleting manifest", cid = manifestCid
+
+    if err =? (await self.blockStore.delBlock(treeCid)).errorOption:
+      error "Error deleting block", cid = treeCid
+
+    if err =? (await self.recordCheckpoint(treeCid, datasetMd)).errorOption:
+      return failure(err)
+
+    return success()
+  else:
+    datasetMd.checkpoint.progress = index
+    return await self.deleteBatch(treeCid, datasetMd)
+
+proc superviseDatasetDeletion(self: DatasetMaintainer, treeCid: Cid, datasetMd: DatasetMetadata): Future[void] {.async.} =
+  logScope:
+    treeCid = treeCid
+    manifestsCids = datasetMd.manifestsCids
+    expiry = datasetMd.expiry
+    leavesCount = datasetMd.leavesCount
+
+  try:
+    if datasetMd.checkpoint.progress == 0 and datasetMd.checkpoint.timestamp == 0:
+      info "Initiating deletion of a dataset"
+    else:
+      info "Retrying deletion of a dataset", progress = datasetMd.checkpoint.progress, timestamp = datasetMd.checkpoint.timestamp
+
+    if err =? (await self.deleteBatch(treeCid, datasetMd)).errorOption:
+      error "Error occurred during deletion of a dataset", msg = err.msg
+    else:
+      info "Dataset deletion complete"
+  except CancelledError as err:
+    raise err
+  except CatchableError as err:
+    error "Unexpected error during dataset deletion", msg = err.msg, treeCid = treeCid
+
+proc listDatasetMetadata*(
+  self: DatasetMaintainer
+): Future[?!AsyncIter[(Cid, DatasetMetadata)]] {.async.} =
+  without queryKey =? createDatasetMetadataQueryKey(), err:
+    return failure(err)
+
+  without queryIter =? await query[DatasetMetadata](self.metaDs, Query.init(queryKey)), err:
```

Comment:
How big is this query going to be, we should be careful and perhaps use pagination if this gets too large.

Reply:
It depends on how many datasets are we storing. I would say that number probably will not ever exceed 1k in practice. But if you want I can add pagination.